### PR TITLE
examples/fluids: Get Viscous flux from solution for flat plate boundary layer problems

### PR DESCRIPTION
**Pull request by jrwrigh**
Use solution gradients to calculate viscous flux in the boundary integral terms.

TODOs:
- [x] Add P_0 to Newtonian context
    - [x] Set new newtonian_ctx P0 in blasius.c
- [x] Move STGStrongInflow QF into newtonian.h
- [x] Move Blasius_Outflow QF into newtonian.h
- [x] Add viscous flux to the inflow and outflow QFs
- [x] Set the inflow and outflow QFs in newtonian.c
- [x] Move State structs and associated helper functions to separate header file
- [x] Pass solution gradient to inflow and outflow QFs
    - [x] Make name `Grad q`
- [x] Make new tests (accounting for change in boundary conditions)

### CODE REVIEW

File: examples/fluids/qfunctions/newtonian.h [hunk 1]
```diff
@@ -15,6 +15,7 @@
 #include <math.h>
 #include <ceed.h>
 #include "newtonian_types.h"
+#include "newtonian_state.h"
```
Comment (jedbrown) [inline examples/fluids/qfunctions/newtonian.h:18]: This commit should delete the now-duplicate code from this file. (It doesn't even compile due to redefinition errors.)

Comment (jrwrigh) [reply examples/fluids/qfunctions/newtonian.h:18]: Yep, I realized that soon after I pushed, but forgot to push the fix up last night. Somehow got dropped after moving my local branch from a custom `jed/fluids-jacobian`+`main` branch to `jed/fluid-jacobian` after you merged in `main`.

**Comment by jrwrigh** [comment]
Assuming tests pass, this should be ready for review/merge.

File: examples/fluids/qfunctions/stg_shur14.h [hunk 2]
```diff
@@ -274,70 +303,79 @@ CEED_QFUNCTION(STGShur14_Inflow)(void *ctx, CeedInt Q,
...
-                            norm[j] * P);
-
-    // -- Total Energy Density
-    v[4][i] -= wdetJb * u_normal * (E + P);
-  }
+    v[0][i] = - wdetJb * drho * u_normal;
+    for (int j=0; j<3; j++)
+      v[j+1][i] = -wdetJb * (drho * u_normal * velocity[j] + norm[j] * dP);
+    v[4][i] = - wdetJb * u_normal * (dE + dP);
+  } // End Quadrature Point Loop
```
Comment (jedbrown) [inline examples/fluids/qfunctions/stg_shur14.h:378]: @LeilaGhaffari Heads up about this incoming code. If we use `FluxInviscid_fwd`, it'll help make it match what you're doing with primitive variables.

File: examples/fluids/src/setuplibceed.c [hunk 3]
```diff
@@ -243,14 +270,20 @@ PetscErrorCode CreateOperatorForDomain(Ceed ceed, DM dm, SimpleBC bc,
...
                              CEED_BASIS_COLLOCATED, q_data_sur);
+        CeedOperatorSetField(op_apply_outflow_jacobian, "x", elem_restr_x_sur,
+                             ceed_data->basis_x_sur, ceed_data->x_coord);
         CeedOperatorSetField(op_apply_outflow_jacobian, "surface jacobian data",
                              elem_restr_jd_i_sur,
                              CEED_BASIS_COLLOCATED, jac_data_sur);
         CeedOperatorSetField(op_apply_outflow_jacobian, "v", elem_restr_q_sur,
                              ceed_data->basis_q_sur, CEED_VECTOR_ACTIVE);
+        CeedOperatorSetField(op_apply_outflow_jacobian, "dummy", elem_restr_q_sur,
+                             ceed_data->basis_q_sur, CEED_VECTOR_ACTIVE);
```
Comment (jedbrown) [inline examples/fluids/src/setuplibceed.c:286]: We can get rid of this now, right?

File: examples/fluids/src/setuplibceed.c [hunk 4]
```diff
@@ -600,12 +639,19 @@ PetscErrorCode SetupLibceed(Ceed ceed, CeedData ceed_data, DM dm, User user,
...
                           q_data_size_sur, CEED_EVAL_NONE);
     CeedQFunctionAddInput(ceed_data->qf_apply_inflow_jacobian, "x", num_comp_x,
                           CEED_EVAL_INTERP);
+    CeedQFunctionAddInput(ceed_data->qf_apply_inflow_jacobian,
+                          "surface jacobian data",
+                          jac_data_size_sur, CEED_EVAL_NONE);
     CeedQFunctionAddOutput(ceed_data->qf_apply_inflow_jacobian, "v", num_comp_q,
                            CEED_EVAL_INTERP);
+    CeedQFunctionAddOutput(ceed_data->qf_apply_inflow_jacobian, "dummy",
+                           num_comp_q*dim_sur, CEED_EVAL_GRAD);
```
Comment (jedbrown) [inline examples/fluids/src/setuplibceed.c:654]: And this?

File: examples/fluids/src/setuplibceed.c [hunk 5]
```diff
@@ -637,13 +685,19 @@ PetscErrorCode SetupLibceed(Ceed ceed, CeedData ceed_data, DM dm, User user,
...
                           q_data_size_sur, CEED_EVAL_NONE);
+    CeedQFunctionAddInput(ceed_data->qf_apply_outflow_jacobian, "x", num_comp_x,
+                          CEED_EVAL_INTERP);
     CeedQFunctionAddInput(ceed_data->qf_apply_outflow_jacobian,
                           "surface jacobian data",
                           jac_data_size_sur, CEED_EVAL_NONE);
     CeedQFunctionAddOutput(ceed_data->qf_apply_outflow_jacobian, "v", num_comp_q,
                            CEED_EVAL_INTERP);
+    CeedQFunctionAddOutput(ceed_data->qf_apply_outflow_jacobian, "dummy",
+                           num_comp_q*dim_sur, CEED_EVAL_GRAD);
```
Comment (jedbrown) [inline examples/fluids/src/setuplibceed.c:700]: +

**Review by jedbrown — APPROVED** [review]
Generally looks good, though I think we should remove the `"dummy"` parts (with https://github.com/CEED/libCEED/pull/988, which should merge momentarily). You could cherry-pick 84e7b512ee4f3ef84f7ca44736ead8e0ffaedab4 to do that.

File: examples/fluids/src/setuplibceed.c (hunk 3 above)
Comment (jedbrown) [reply examples/fluids/src/setuplibceed.c:286]: #988 is merged so I'd suggest just cherry-picking the above. Are you planning on a squash-merge or doing some light squashing in this branch's history?

**Comment by jedbrown** [comment]
Do you want this squash-merged or normal merged?

**Comment by jrwrigh** [comment]
Sorry I missed this. I'm ok with either. I can do a quick bit of history clean up if we want it normal merged.

Also, am I "allowed" to merge it in myself? I know I have write access, but still a bit apprehensive of doing it. It's more power than I'm used to wielding. haha
